pkg/identity/cache/allocator_utils.go

```diff
@@ -0,0 +1,159 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package cache
+
+import (
+	"context"
+
+	"github.com/cilium/cilium/pkg/allocator"
+	"github.com/cilium/cilium/pkg/identity"
+	identitymodel "github.com/cilium/cilium/pkg/identity/model"
+	"github.com/cilium/cilium/pkg/labels"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/metrics"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/sirupsen/logrus"
+)
+
+type UpdateIDsFunc func(added, deleted IdentityCache)
+type ReleaseFunc func(ctx context.Context, id *identity.Identity, notifyOwner bool) (released bool, err error)
+type GetLocalIDsFunc func() map[identity.NumericIdentity]*identity.Identity
+
+var UnknownIdentity = identity.NewIdentity(identity.IdentityUnknown, labels.Labels{labels.IDNameUnknown: labels.NewLabel(labels.IDNameUnknown, "", labels.LabelSourceReserved)})
+
+func WithholdLocalIdentities(localIDs, localNodeIDs *LocalIdentityCache, nids []identity.NumericIdentity) {
+	log.WithField(logfields.Identity, nids).Debug("Withholding numeric identities for later restoration")
+
+	// The allocators will return any identities that are not in-scope.
+	nids = localIDs.withhold(nids)
+	nids = localNodeIDs.withhold(nids)
+	if len(nids) > 0 {
+		log.WithField(logfields.Identity, nids).Error("Attempt to restore invalid numeric identities.")
+	}
+}
+
+func UnwithholdLocalIdentities(localIDs, localNodeIDs *LocalIdentityCache, nids []identity.NumericIdentity) {
```

Comment:
Also, is there a reason not to use composition to solve this sort of problem, but rather add these static functions?

Reply:
You're right. At first it seemed simpler to do it like this, but with more consideration I was able to extract the common code for local identity allocation into a new struct `LocalCacheAllocator`, which is used by the `CachingIdentityAllocator` to allocate reserved and local identities, before executing it's logic for allocating global identities. The same will be used later by the `LocalOnlyCachingIDAllocator`.